### PR TITLE
planner: Fix SMJ hint, support SMJ with descending order.

### PR DESCRIPTION
**Pull request by ichn-hu**
### What problem does this PR solve? 

This PR fixes https://github.com/pingcap/tidb/issues/14483

### What is changed and how it works?

Do not create SMJ physical plan if PhysicalProperty has desc equals to true.

### Check List 

Tests 

 - Unit test

Side effects

 - Possible performance regression

Release note

 - Fixing a correctness bug in execution, need to be cherry-picked to release version.

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -199,7 +199,7 @@ func (p *LogicalJoin) getEnforcedMergeJoin(prop *property.PhysicalProperty) []Ph
 	// Check whether SMJ can satisfy the required property
 	offsets := make([]int, 0, len(p.LeftJoinKeys))
 	all, desc := prop.AllSameOrder()
-	if !all {
+	if !all || desc {
```
Comment (lzmhhh123) [inline planner/core/exhaust_physical_plans.go]: Why ban the descent merge join plan? It seems to change it in the executor is not hard. Only keep similar with index merge join is ok.

Comment (ichn-hu) [reply planner/core/exhaust_physical_plans.go]: Perhaps we want a quick fix?

Comment (francis0407) [reply planner/core/exhaust_physical_plans.go]: We just release v3.0.9, and there is still time for fixing th executor before v3.0.10. Can we just fix the root cause?

Comment (ichn-hu) [reply planner/core/exhaust_physical_plans.go]: Fair enough, I'll fix it soon.

**Comment by ichn-hu** [comment]
@lzmhhh123 @francis0407 PTAL

File: executor/merge_join_test.go [hunk 2]
```diff
@@ -395,6 +395,23 @@ func (s *testSuite2) TestMergeJoin(c *C) {
...
+	tk.MustExec("create table t (a int, key(a))")
+	tk.MustExec("create table t1 (a int, key(a))")
+	tk.MustExec("insert into t values (1), (2), (3)")
+	tk.MustExec("insert into t1 values (1), (2), (3)")
+	tk.MustQuery("select /*+ TIDB_SMJ(t1, t2) */ t.a from t, t1 where t.a = t1.a order by t1.a desc").Check(testkit.Rows(
+		"3", "2", "1"))
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int, key(a), key(b))")
+	tk.MustExec("insert into t values (1,1),(1,2),(1,3),(2,1),(2,2),(3,1),(3,2),(3,3)")
+	tk.MustQuery("select /*+ TIDB_SMJ(t1, t2) */ t1.a from t t1, t t2 where t1.a = t2.b order by t1.a desc").Check(testkit.Rows(
```
Comment (SunRunAway) [inline executor/merge_join_test.go]: The test case does not cover `getMergeJoin`.

**Comment by zz-jason** [comment]
you can use [`/merge`](https://github.com/pingcap/community/blob/master/contributors/command-help.md#run-auto-mergemerge) command instead of remove and add `status/can merge` label 😂

**Comment by ichn-hu** [comment]
@zz-jason PTAL again [xiaoku]

**Comment by ichn-hu** [comment]
It's wield, it fails UT every time with different test cases all seemingly not related to the modification I've made... This time it is https://internal.pingcap.net/idc-jenkins/blue/rest/organizations/jenkins/pipelines/tidb_ghpr_unit_test/runs/22727/nodes/67/log/?start=0

**Pull request merged**